### ISSUE TITLE
Running the model for only test, without annotation

### ISSUE THREAD

**Issue by alpertungakin** [likaiucas/OBM#2]
Hi. I wish to run the model for only test mode. I have no annotations for my images. I have made the following changes to the model script and created an inference script, but it returns as a 0-sized tensor. Do you have any different suggestions? I think the following command does not work well because it needs annotation files.
`python tools/test_offset.py --config configs/obm_seg/obm_seg_b.py`

inference script:
```python
from mmdet.apis import init_detector, inference_detector
from mmdet.apis import show_result_pyplot
import mmcv

# Define the model configuration file and checkpoint path
config_file = '/home/alper-tunga/Desktop/OBM-main/configs/obm_seg/obm_seg_b.py'
checkpoint_file = '/home/alper-tunga/Desktop/OBM-main/trainedmodel/epoch_48.pth'

# Initialize the model
model = init_detector(config_file, checkpoint_file, device='cuda:0')  # Use 'cpu' if no GPU is available

# Path to the test image folder
test_image_folder = '/home/alper-tunga/Desktop/OBM-main/test-images/'

img_path = test_image_folder + 'input_image_0_3.tif'
result = inference_detector(model, img_path)

# Unpack the result
bboxes, labels, masks = result

# Visualize the result
show_result_pyplot(model, img_path, (bboxes, labels))

# Perform inference on all images in the folder
results = []
for img_path in mmcv.scandir(test_image_folder):
full_img_path = f"{test_image_folder}/{img_path}"
result = inference_detector(model, full_img_path)
results.append((img_path, result))

print(result)
```
changed model script. I made the changes on _forward_test_ function (obm_seg.py):
```python
from mmcv.runner import auto_fp16
import torch
from ..builder import DETECTORS, build_backbone, build_head, build_neck
from torch import nn
from torch.nn import functional as F
import random
from mmdet.models.backbones import ImageEncoderViT
from mmdet.models.dense_heads import PromptEncoder
from mmdet.utils import get_root_logger
from typing import Any, Dict, List, Tuple
# Replace relative imports with absolute imports
from ..builder import DETECTORS, build_backbone, build_head, build_neck
from mmdet.models.backbones import ImageEncoderViT
from mmdet.models.dense_heads import PromptEncoder
from mmdet.models.roi_heads import OffsetDecoder
from mmdet.models.detectors.base import BaseDetector
from mmdet.core import build_sampler
from copy import deepcopy

def noise(x, mean=0):
    # offset = torch.tensor(torch.rand(x.shape).cuda()*mean, dtype=torch.int)
    offset = torch.rand(x.shape).cuda()*mean
    x[:, :2]-=offset[:, :2]
    x[:, 2:4]+=offset[:, 2:4]
    x[x<0]=0
    x[x>1024]=1024.
    return x

@DETECTORS.register_module()
class obm_seg(BaseDetector):
    mask_threshold: float = 0.0
    image_format: str = "RGB"

    # __init__ 方法:
    # 1. 输入参数:
    #     - image_encoder: 用于将图像编码为图像 embedding 的骨干网络,用于有效地预测掩码。
    #     - prompt_encoder: 用于编码各种类型的输入提示。
    #     - mask_decoder: 根据图像 embedding 和编码的提示预测掩码。
    #     - pixel_mean: 输入图像像素的归一化均值。
    #     - pixel_std: 输入图像像素的归一化标准差。
    # 2. 调用父类初始化。
    # 3. 记录 image_encoder、prompt_encoder 和 mask_decoder。
    # 4. 使用 register_buffer 注册 pixel_mean 和 pixel_std。
    def __init__(
        self,
        image_encoder: ImageEncoderViT,
        prompt_encoder: PromptEncoder,
        mask_decoder: OffsetDecoder,
        pretrained=None,
        train_cfg=None,
        test_cfg=None,
        sampler=None,
        pixel_mean = [0, 0, 0], 
        pixel_std = [1, 1, 1],
        noise_box=None,
    ) :
        self.train_cfg = train_cfg
        self.test_cfg = test_cfg
        self.sampler = build_sampler(sampler, context=self) if sampler is not None else None
        """-> None
        SAM predicts object masks from an image and input prompts.

        Arguments:
          image_encoder (ImageEncoderViT): The backbone used to encode the
            image into image embeddings that allow for efficient mask prediction.
          prompt_encoder (PromptEncoder): Encodes various types of input prompts.
          mask_decoder (MaskDecoder): Predicts masks from the image embeddings
            and encoded prompts.
          pixel_mean (list(float)): Mean values for normalizing pixels in the input image.
          pixel_std (list(float)): Std values for normalizing pixels in the input image.
        """
        super().__init__()
        self.noise_box = noise_box
        if isinstance(image_encoder, nn.Module):
            self.image_encoder = image_encoder
        else:
            self.image_encoder = build_backbone(image_encoder)
        if isinstance(prompt_encoder, nn.Module):
            self.prompt_encoder = prompt_encoder
        else:
            self.prompt_encoder =build_head(prompt_encoder)
        if isinstance(mask_decoder, nn.Module):
            self.mask_decoder = mask_decoder
        else:
            self.mask_decoder = build_head(mask_decoder)
        self.register_buffer("pixel_mean", torch.Tensor(pixel_mean).view(-1, 1, 1), False)
        self.register_buffer("pixel_std", torch.Tensor(pixel_std).view(-1, 1, 1), False)

    # 这个 device 属性返回 pixel_mean 的设备类型。
    # pixel_mean 在 __init__ 方法中使用 register_buffer 注册为 buffer, 它的设备类型由输入图像的设备类型决定。
    # 所以,这个 device 属性返回模型中用于图像处理的设备类型,通常为CPU或GPU。
    @property
    def device(self) :#-> Any
        return self.pixel_mean.device
    def forward_train(self,
                      img,
                      img_metas,
                      gt_bboxes=None,
                      gt_labels=None,
                      gt_bboxes_ignore=None,
                      gt_building_masks=None,
                      gt_masks=None,
                      gt_offsets=None,
                      **kwargs):
        assert gt_masks or gt_bboxes
        if self.sampler is not None:
            prompt_len= len(gt_offsets[0])
            if prompt_len>self.train_cfg['max_num']:
                loc = self.sampler.custom_sampler(gt_offsets[0]).cpu().numpy()
                gt_bboxes[0] = gt_bboxes[0][loc]
                gt_labels[0] = gt_labels[0][loc]
                gt_masks[0] = gt_masks[0][loc]
                if gt_building_masks is not None:
                    gt_building_masks[0] =  gt_building_masks[0][loc]#gt_seg[0][loc,:]
                else:
                    gt_building_masks = gt_masks
                gt_offsets[0] = gt_offsets[0][loc]
        
        prompt_len= len(gt_labels[0])
        if prompt_len>self.train_cfg['max_num']:
            loc = random.sample(range(0, prompt_len), self.train_cfg['max_num'])
            gt_bboxes[0] = gt_bboxes[0][loc,:]
            gt_labels[0] = gt_labels[0][loc]
            gt_masks[0] = gt_masks[0][loc,:]
            if gt_building_masks is not None:
                gt_building_masks[0] =  gt_building_masks[0][loc,:]#gt_seg[0][loc,:]
            else:
                gt_building_masks = gt_masks
            gt_offsets[0] = gt_offsets[0][loc,:]
        features = self.image_encoder(img)
        losses = dict()
        sparse_embeddings, dense_embeddings = self.prompt_encoder(
                points=None,
                boxes=gt_bboxes[0],
                masks=None,
            )
        loss = self.mask_decoder.forward_train(
                image_embeddings=features,
                image_pe=self.prompt_encoder.get_dense_pe(),
                sparse_prompt_embeddings=sparse_embeddings,
                dense_prompt_embeddings=dense_embeddings,
                multimask_output=False,
                gt_offsets=gt_offsets,
                gt_masks=gt_masks,
                gt_building_masks=gt_building_masks,
                **kwargs
            )
        losses.update(loss)
        return losses
# prompt boxes        
    def forward_test(self, 
                    img,
                    img_metas,
                    gt_bboxes=None,
                    gt_labels=None,
                    gt_bboxes_ignore=None,
                    gt_masks=None,
                    proposals=None,
                    **kwargs):
        # Handle the case where gt_bboxes is None
        if gt_bboxes is None:
            # Use the device of the first image in the batch
            device = img[0].device
            gt_bboxes = [[torch.zeros((1, 4), device=device)]]  # Default dummy boxes

        return_box = deepcopy(gt_bboxes)

        if self.noise_box is not None:
            gt_bboxes = [noise(gt_bboxes[0][0], self.noise_box).unsqueeze(0)]

        x = self.image_encoder(img[0])
        sparse_embeddings, dense_embeddings = self.prompt_encoder(
            points=None,
            boxes=gt_bboxes[0][0],  # Use the (possibly modified) gt_bboxes
            masks=None,
        )
        offset, prob, masks = self.mask_decoder.forward_test(
            image_embeddings=x,
            image_pe=self.prompt_encoder.get_dense_pe(),
            sparse_prompt_embeddings=sparse_embeddings,
            dense_prompt_embeddings=dense_embeddings,
        )
        masks = self.postprocess_masks(masks, (1024, 1024), (1024, 1024))
        mean_first_channels = masks[:, :2, :, :].mean(dim=1, keepdim=True)
        mean_last_channels = masks[:, 2:, :, :].mean(dim=1, keepdim=True)
        masks = torch.cat([mean_first_channels, mean_last_channels], dim=1)
        masks = masks > self.mask_threshold

        # Generate dummy labels if not provided
        if gt_labels is None:
            gt_labels = [torch.zeros((gt_bboxes[0][0].shape[0],), dtype=torch.int64, device=gt_bboxes[0][0].device)]

        # Return bboxes, labels, and masks for visualization
        return return_box[0][0], gt_labels[0], masks
    def postprocess_masks(
        self,
        masks: torch.Tensor,
        input_size: Tuple[int, ...],
        original_size: Tuple[int, ...],
    ) :
        """-> torch.Tensor
        Remove padding and upscale masks to the original image size.

        Arguments:
          masks (torch.Tensor): Batched masks from the mask_decoder,
            in BxCxHxW format.
          input_size (tuple(int, int)): The size of the image input to the
            model, in (H, W) format. Used to remove padding.
          original_size (tuple(int, int)): The original size of the image
            before resizing for input to the model, in (H, W) format.

        Returns:
          (torch.Tensor): Batched masks in BxCxHxW format, where (H, W)
            is given by original_size.
        """
        masks = F.interpolate(
            masks,
            (self.image_encoder.img_size, self.image_encoder.img_size),
            mode="bilinear",
            align_corners=False,
        )
        masks = masks[..., : input_size[0], : input_size[1]]
        masks = F.interpolate(masks, original_size, mode="bilinear", align_corners=False)
        return masks
    @auto_fp16(apply_to=('img', ))
    def forward(self, img, img_metas, return_loss=True, **kwargs):
        """Calls either :func:`forward_train` or :func:`forward_test` depending
        on whether ``return_loss`` is ``True``.

        Note this setting will change the expected inputs. When
        ``return_loss=True``, img and img_meta are single-nested (i.e. Tensor
        and List[dict]), and when ``resturn_loss=False``, img and img_meta
        should be double nested (i.e.  List[Tensor], List[List[dict]]), with
        the outer list indicating test time augmentations.
        """
        if torch.onnx.is_in_onnx_export():
            assert len(img_metas) == 1
            return self.onnx_export(img[0], img_metas[0])

        if return_loss:
            return self.forward_train(img, img_metas, **kwargs)
        else:
            return self.forward_test(img, img_metas, **kwargs)
        
    def aug_test(self,):
        pass
    def extract_feat(self):
        pass
    def simple_test(self):
        pass
```
Dataset script. I have made the changes on 'test' key. (obm_bonai_instance_prompt.py):
```python
dataset_type = 'BONAI'
data_root = '/test-images/'
img_norm_cfg = dict(
    mean=[123.675, 116.28, 103.53], std=[58.395, 57.12, 57.375], to_rgb=True)
train_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(type='LoadAnnotations', 
         with_bbox=True,
         with_mask=True,
         with_offset=True),
    dict(type='RandomFlip', flip_ratio=0.5, direction=['horizontal', 'vertical']),
    dict(type='Resize', img_scale=(1024,1024), keep_ratio=True),
    dict(type='Normalize', **img_norm_cfg),
    dict(type='Pad', size_divisor=32),
    dict(type='DefaultFormatBundle'),
    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels', 'gt_masks', 'gt_offsets']),
]

test_pipeline = [
    dict(type='LoadImageFromFile'),
    dict(
        type='MultiScaleFlipAug',
        img_scale=(1024, 1024),
        flip=False,  # Set to False if you don't want to flip during inference
        transforms=[
            dict(type='Resize', keep_ratio=True),
            dict(type='Normalize', **img_norm_cfg),
            dict(type='Pad', size_divisor=32),
            dict(type='ImageToTensor', keys=['img']),
            dict(type='Collect', keys=['img']),  # Only collect 'img' for inference
        ])
]

cities = ['shanghai', 'beijing', 'jinan', 'haerbin', 'chengdu']
# cities = ['shanghai', 'beijing', 'haerbin', 'chengdu']

train_ann_file = []
img_prefix = []

val_ann_file = []
val_image_prefix = []
for city in cities:
    train_ann_file.append(data_root + 'coco/bonai_{}_trainval.json'.format(city))
    img_prefix.append("/config_data/BONAI_data/trainval-001/trainval/images")

val_ann_file.append(data_root + 'coco/bonai_jinan_trainval.json')
val_image_prefix.append("/config_data/BONAI_data/trainval-001/trainval/images")
data = dict(
    samples_per_gpu=1,
    workers_per_gpu=1,
    train=dict(
        type=dataset_type,
        ann_file=train_ann_file,
        img_prefix=img_prefix,
        bbox_type='building',
        mask_type='roof',
        pipeline=train_pipeline),
    val=dict(
        type=dataset_type,
        ann_file=val_ann_file[0],
        img_prefix=val_image_prefix[0],
        gt_footprint_csv_file="",
        bbox_type='building',
        mask_type='roof',
        pipeline=test_pipeline),
    test=dict(
        type=dataset_type,
        ann_file=None,  # No annotations needed for inference
        img_prefix='/home/alper-tunga/Desktop/OBM-main/test-images/',
        pipeline=test_pipeline),
)
evaluation = dict(interval=1, metric=['bbox', 'segm'])
`

**Comment by likaiucas** [comment]
For you didn't input any prompts with the image, you get a 0-sized tensor.

Unfortunately, the released model, OBM, only supports extractions with box prompts. Although I proposed another model named PolyFootNet, which supports the automatic and prompt-driven extraction of polygonal results, the paper is under review by TGRS. 
The arxiv version is available at https://arxiv.org/abs/2408.08645, but I did not upload the newest version. 
For inferencing, you should use the configure file _configs/obm_seg_fintune/smlcdr_obm_pretrain.py_, which was trained with ROAM. _configs/obm_seg/obm_seg_b.py_ is used for ablation study. It is OK to use _configs/obm_seg/obm_seg_b.py_, but it will exclude the operation of ROAM.

BTW, to better help you with the BFE problem, I would like to know if you would like to continue my model for scientific research or for technical applications.

**Comment by alpertungakin** [comment]
Thank you for your support. I will use the model for a middle process of my PhD research. I have a bad-shaped orthomosaic so, I need to apply a kind of self-rectification to eliminate the visible façades. Because I will prepare a data stack with other layers but I have a registiration problem now. I divided the mosaic into grids but I do not have bbox annotations for them. This is why I need to use without an annotation-needed test configuration.

**Comment by likaiucas** [comment]
1. What is the resolution of your image? Around 0.5m or 0.2m? 

2. And the distribution of buildings? Crowded or Sparse?

3. Could you show me some cases of your _bad-shaped orthomosaic_?

As there were some tricks and unpublished ideas that I have not included in my latest publication, you can answer the questions mentioned above, and if it is good for you, I will let you know.

**Comment by alpertungakin** [comment]
I will send you an e-mail with the answers. I will send it to [likai211@mails.ucas.ac.cn] if it is OK for you.

**Comment by likaiucas** [comment]
kaili37-c@my.cityu.edu.hk

**Comment by 0mil** [comment]
Hello, @alpertungakin, @likaiucas 

I’ve been trying to follow you two guys environment setup and test-related instructions step-by-step. Thank you for sharing the details — they’ve been very helpful as a reference.

Like @alpertungakin , I’m eager to see this model’s results for my own personal project. However, in my current environment (GPU: `RTX 5090`, WSL2 `Ubuntu-22.04`), it feels almost impossible to get everything fully set up. I’ve already spent whole three days troubleshooting for just setup without success.

Were you eventually able to complete the process and run the project to obtain results?
If so, could you share any advice or steps that helped you get there? It would be greatly appreciated.

Thanks in advance!

**Comment by likaiucas** [comment]
Hi,


I guess its problem with your WSL2. Coz if my memory right, WSL2 with nvidia only supports limited versions of pytorch, which maybe uncompactable. To run it, maybe you need a Docker Image to create a similar environment as we provide, or create an image and manually install it. It canbe established directly on your windows system.&nbsp;


Otherwise, I recommend you to run another github project: Advance_BONAI, where I try to reproduce the BONAI and our methods on a more advanced mmdet and environment. Reminds: although I have fixed almost all bugs, using advancebonai for trainning has a performance gap. If you find some difference with our OBM, you should copy some files from project OBM to Advance_BONAI. I guess this way maybe more difficult. But if you do so, I encourage you to join the maintance of advance bonai project.


BTW, I have another idea to create a supportable environment, which may work.
Use a more latest project's environment, and then use its environment to run our code.&nbsp; https://github.com/opendatalab/MLS-BRN.git


MLSBRN is a project also based on BONAI. So it may work.&nbsp;




Conclusion: I guess my first advice maybe more easy and reliable for you. WSL2 and WSL both are not that popular for doing research, since most of packages need a special edition. However, it is impossible to require the backup providers to create a wls supported environment for those old projects.



---Original---
From: "Jin Hyeong ***@***.***&gt;
Date: Fri, Aug 15, 2025 00:37 AM
To: ***@***.***&gt;;
Cc: "LI ***@***.******@***.***&gt;;
Subject: Re: [likaiucas/OBM] Running the model for only test, withoutannotation (Issue #2)


0mil left a comment (likaiucas/OBM#2)
 
Hello, @alpertungakin, @likaiucas
 
I’ve been trying to follow you two guys environment setup and test-related instructions step-by-step. Thank you for sharing the details — they’ve been very helpful as a reference.
 
Like @alpertungakin , I’m eager to see this model’s results for my own personal project. However, in my current environment (GPU: RTX 5090, WSL2 Ubuntu-22.04), it feels almost impossible to get everything fully set up. I’ve already spent whole three days troubleshooting for just setup without success.
 
Were you eventually able to complete the process and run the project to obtain results?
 If so, could you share any advice or steps that helped you get there? It would be greatly appreciated.
 
Thanks in advance!

**Comment by 0mil** [comment]
Thank you for your advice through and through. I'm gonna try your suggestions right away.
And your work is truly impressive, and I’m looking forward to seeing more of your works. 👍
